From 1f1cdfbf96d55f0603b19fbf95b43a20377d993d Mon Sep 17 00:00:00 2001
From: Valery Zabrovsky <valthebrewer@yandex.ru>
Date: Mon, 24 Nov 2025 09:53:53 +0300
Subject: [PATCH 3/3] Pass random seed as parameter

---
 09_PatchDiff/maze.c | 11 +++++------
 1 file changed, 5 insertions(+), 6 deletions(-)

diff --git a/09_PatchDiff/maze.c b/09_PatchDiff/maze.c
index 732d8cd..9b6e2d8 100644
--- a/09_PatchDiff/maze.c
+++ b/09_PatchDiff/maze.c
@@ -1,6 +1,5 @@
 #include <stdio.h>
 #include <stdlib.h>
-#include <time.h>
 
 char **
 generate_maze(int size, char pass, char wall)
@@ -118,20 +117,20 @@ main(int argc, char *argv[])
     int i;
     int size;
 
-    if (argc < 3) {
+    if (argc < 4) {
         return 1;
     }
 
-    if (argv[1][0] == '\0' || argv[1][1] == '\0') {
+    if (argv[2][0] == '\0' || argv[2][1] == '\0') {
         return 1;
     }
 
-    if ((size = atoi(argv[2])) <= 0) {
+    if ((size = atoi(argv[3])) <= 0) {
         return 1;
     }
 
-    srand(time(NULL));
-    maze = generate_maze(size, argv[1][0], argv[1][1]);
+    srand(atoi(argv[1]));
+    maze = generate_maze(size, argv[2][0], argv[2][1]);
 
     for (i = 0; i <= 2 * size; ++i) {
         printf("%.*s\n", 2 * size + 1, maze[i]);
-- 
2.51.1

